### PR TITLE
Impl ASSERT_END_OF_BYTECODE_SEGMENTS hint

### PR DESCRIPTION
**Pull request by notlesh**
Implements the `ASSERT_END_OF_BYTECODE_SEGMENTS` hint.

Issue Number: N/A

## Type

- [x] feature
- [ ] bugfix
- [ ] dev (no functional changes, no API changes)
- [ ] fmt (formatting, renaming)
- [ ] build
- [ ] docs
- [ ] testing

## Breaking changes?

- [ ] yes
- [x] no

### CODE REVIEW

File: src/hints/compiled_class.rs [hunk 1]
```diff
@@ -30,3 +30,22 @@ pub fn assign_bytecode_segments(
...
+    assert next(bytecode_segments, None) is None"#
+};
+pub fn assert_end_of_bytecode_segments(
+    _vm: &mut VirtualMachine,
+    exec_scopes: &mut ExecutionScopes,
+    _ids_data: &HashMap<String, HintReference>,
+    _ap_tracking: &ApTracking,
+    _constants: &HashMap<String, Felt252>,
+) -> Result<(), HintError> {
+    let mut bytecode_segments: <Vec<BytecodeSegment> as IntoIterator>::IntoIter = exec_scopes.get(vars::scopes::BYTECODE_SEGMENTS)?;
```
Comment (notlesh) [inline src/hints/compiled_class.rs]: This type scares me... A test that uses both this hint and the one above (`ASSIGN_BYTECODE_SEGMENTS`) would help.

File: src/hints/compiled_class.rs [hunk 2]
```diff
@@ -30,3 +30,73 @@ pub fn assign_bytecode_segments(
...
+    };
+    use crate::starkware_utils::commitment_tree::base_types::Length;
+
+    #[test]
+    fn test_bytecode_segment_hints() {
+        // tests both ASSIGN_BYTECODE_SEGMENTS and ASSERT_END_OF_BYTECODE_SEGMENTS. The first
+        // should prepare an iterater and put it in ExecutionScopes, the second should ensure that
+        // the iterator is exhausted.
+
+        let mut vm = VirtualMachine::new(false);
```
Comment (maciejka) [inline src/hints/compiled_class.rs:69]: Default vm setup deserves a fixture. I mean all hint arguments: vm,  exec_scopes, ...

Comment (notlesh) [reply src/hints/compiled_class.rs:69]: I like that idea, especially with the magic values (like needing 2 memory segments and fp to be 2 or more).

By the way, @odesenfans cleared up the two mem segments you asked about earlier:
* Program code executes at segment 0
* Variables are in segment 1
* The VM does not create these segments on its own

I'd support a fixture for this, but I think it should go in a new PR (maybe after we finish up the last few hints).

File: src/hints/compiled_class.rs (hunk 1 above)
Comment (notlesh) [reply src/hints/compiled_class.rs]: Test added in the last few commits

**Review by maciejka — APPROVED** [review]

**Review by maciejka — COMMENTED** [review]
Not on critical path.

**Review by odesenfans — APPROVED** [review]

**Pull request merged**